### PR TITLE
Use DoH from the OS DNS provider OR a fallback DoH provider

### PR DESCRIPTION
**Pull request by arthuredelstein**
This behavior is behind a flag: BraveFallbackDoHProvider, and involves an integer parameter, kBraveFallbackDoHProviderEndpointIndex:
1 = Quad9
2 = Wikimedia
3 = Cloudflare

Resolves https://github.com/brave/brave-browser/issues/35677

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: chromium_src/net/dns/dns_client.cc [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+namespace net {
+namespace {
+
+struct fallbackEndpoint {
+  const char* name;
+  const char* address;
+};
+
+constexpr fallbackEndpoint endpoints[] = {
```
Comment (goodov) [inline chromium_src/net/dns/dns_client.cc:21]: pls consider:
1. adding `enum class DohFallbackEndpointType { ... };` and using it to create `DohFallbackEndpoint` struct that you then export as `base::span<const DohFallbackEndpoint> GetDohFallbackEndpoints();` via `net/dns/public/secure_dns_mode.h` header.
2. using `-1` as a default value for the feature and handle it accordingly in UI/Android codebase.

don't use strings as identifiers, don't use 0 as a default value.

Comment (arthuredelstein) [reply chromium_src/net/dns/dns_client.cc]: I ended up using an enum class as you suggested, which brings some helpful type checking in C++. But it does result in using strings as identifiers in Java. Interested in your thoughts on this approach.

File: browser/about_flags.cc [hunk 2]
```diff
@@ -919,6 +919,14 @@
           kOsAll,                                                              \
           FEATURE_VALUE_TYPE(net::features::kBraveHttpsByDefault),             \
       },                                                                       \
+      {                                                                        \
+          "fallback-dns-over-https",                                           \
+          "Use a fallback DoH provider",                                       \
+          "In Automatic DoH mode, use a fallback DoH provider if the current " \
+          "provider doesn't offer Secure DNS.",                                \
+          kOsAll,                                                              \
+          FEATURE_VALUE_TYPE(net::features::kBraveFallbackDoHProvider),        \
```
Comment (goodov) [inline browser/about_flags.cc:928]: consider exposing this via `FEATURE_WITH_PARAMS_VALUE_TYPE`.

File: chromium_src/net/dns/dns_client.cc (hunk 1 above)
Comment (goodov) [reply chromium_src/net/dns/dns_client.cc]: > But it does result in using strings as identifiers in Java. Interested in your thoughts on this approach.

IMO this sounds better than using arbitrary integers for feature parameters.

**Review by mkarolin — APPROVED** [review]
`strings`++

File: chromium_src/net/dns/dns_client.cc [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+struct DohFallbackEndpoint {
+  const char* name;
+  const char* address;
+};
+
+constexpr auto kDohFallbackEndpoints{
+    base::MakeFixedFlatMap<DohFallbackEndpointType, DohFallbackEndpoint>({
+        {DohFallbackEndpointType::kQuad9,
+         {"Quad9", "https://dns.quad9.net/dns-query"}},
```
[the remark this answers is not in the document]
Comment (arthuredelstein) [reply chromium_src/net/dns/dns_client.cc:26]: You are right; fixed.

File: chromium_src/net/dns/dns_client.cc [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        {DohFallbackEndpointType::kWikimedia,
+         {"Wikimedia", "https://wikimedia-dns.org/dns-query"}},
+        {DohFallbackEndpointType::kCloudflare,
+         {"Cloudflare", "https://cloudflare-dns.com/dns-query"}},
+    })};
+
+std::vector<DnsOverHttpsServerConfig> MaybeAddFallbackDohServer(
+    const std::vector<DnsOverHttpsServerConfig>& doh_servers) {
```
[the remark this answers is not in the document]
Comment (arthuredelstein) [reply chromium_src/net/dns/dns_client.cc:34]: done

File: chromium_src/chrome/browser/ui/webui/settings/shared_settings_localized_strings_provider.cc [hunk 5]
```diff
@@ -36,27 +38,53 @@ bool ShouldReplaceSecureDNSDisabledDescription() {
...
-       IDS_SETTINGS_SECURE_DNS_DISABLED_BY_BRAVE_VPN}};
-
-  html_source->AddLocalizedStrings(kLocalizedStrings);
+  if (base::FeatureList::IsEnabled(net::features::kBraveFallbackDoHProvider)) {
+    static const net::DohFallbackEndpointType endpoint =
+        net::features::kBraveFallbackDoHProviderEndpoint.Get();
+    if (endpoint != net::DohFallbackEndpointType::kNone) {
+      static webui::LocalizedString kLocalizedStrings[] = {
+          kAlternateLocalizedStrings.at(endpoint)};
+      html_source->AddLocalizedStrings(kLocalizedStrings);
```
[the remark this answers is not in the document]
Comment (arthuredelstein) [reply chromium_src/chrome/browser/ui/webui/settings/shared_settings_localized_strings_provider.cc:76]: Good catch; fixed. Thanks!

File: chromium_src/chrome/browser/ui/webui/settings/shared_settings_localized_strings_provider.cc [hunk 6]
```diff
@@ -36,27 +38,53 @@ bool ShouldReplaceSecureDNSDisabledDescription() {
...
+    }
+  }
+#if BUILDFLAG(ENABLE_BRAVE_VPN) && BUILDFLAG(IS_WIN)
+  if (ShouldReplaceSecureDNSDisabledDescription()) {
+    static constexpr webui::LocalizedString kLocalizedStrings[] = {
+        {"secureDnsDisabledForManagedEnvironment",
+         IDS_SETTINGS_SECURE_DNS_DISABLED_BY_BRAVE_VPN}};
+    html_source->AddLocalizedStrings(kLocalizedStrings);
```
[the remark this answers is not in the document]
Comment (arthuredelstein) [reply chromium_src/chrome/browser/ui/webui/settings/shared_settings_localized_strings_provider.cc:84]: fixed

**Review by goodov — APPROVED** [review]

**Comment by arthuredelstein** [comment]
Thanks for the reviews!

File: chromium_src/net/dns/dns_client.cc [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    base::MakeFixedFlatMap<DohFallbackEndpointType, const char*>({
+        {DohFallbackEndpointType::kQuad9, "https://dns.quad9.net/dns-query"},
+        {DohFallbackEndpointType::kWikimedia,
+         "https://wikimedia-dns.org/dns-query"},
+        {DohFallbackEndpointType::kCloudflare,
+         "https://cloudflare-dns.com/dns-query"},
+    })};
+
+std::vector<DnsOverHttpsServerConfig> MaybeAddFallbackDohServer(
+    const std::vector<DnsOverHttpsServerConfig> doh_servers) {
```
Comment (goodov) [inline chromium_src/net/dns/dns_client.cc:28]: wait, why is it `const`? this way the value is forced to be copied on `return doh_servers;` statements.

File: chromium_src/net/dns/dns_client.cc [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+      net::features::kBraveFallbackDoHProviderEndpoint.Get();
+  if (endpoint == DohFallbackEndpointType::kNone) {
+    return doh_servers;
+  }
+  const char* endpointAddress = kDohFallbackEndpointAddresses.at(endpoint);
+  auto fallbackDohServer =
+      DnsOverHttpsServerConfig::FromString(endpointAddress);
+  std::vector<DnsOverHttpsServerConfig> extended_doh_servers = doh_servers;
+  if (fallbackDohServer.has_value()) {
+    extended_doh_servers.push_back(fallbackDohServer.value());
```
Comment (goodov) [inline chromium_src/net/dns/dns_client.cc:42]: remove `extended_doh_servers ` and just append to `doh_servers`.

File: chromium_src/net/dns/dns_client.cc (hunk 7 above)
Comment (arthuredelstein) [reply chromium_src/net/dns/dns_client.cc:28]: Ah, you are right. I will follow up with a fix.

File: chromium_src/net/dns/dns_client.cc (hunk 8 above)
Comment (arthuredelstein) [reply chromium_src/net/dns/dns_client.cc:42]: Also fixed in https://github.com/brave/brave-core/pull/22078

File: chromium_src/net/dns/dns_client.cc (hunk 7 above)
Comment (arthuredelstein) [reply chromium_src/net/dns/dns_client.cc:28]: https://github.com/brave/brave-core/pull/22078

**Pull request merged**